fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,357 @@
+    DB: Blockstore + Clone + 'static + Send,
+{
+    /// Creates a new V2Snapshot struct. Caller ensure store
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        let state_tree =
+            StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_params.state_root)?;
+        let state_tree_root = state_params.state_root;
+
+        let block_state_params_bytes = fvm_ipld_encoding::to_vec(&(state_params, block_height))?;
+        let block_state_params_cid =
+            Cid::new_v1(DAG_CBOR, Code::Blake2b256.digest(&block_state_params_bytes));
+
+        Ok(Self {
+            state_tree_root,
+            state_tree,
+            block_state_params_bytes,
+            block_state_params_cid,
+        })
+    }
+
+    async fn read_car(file: tokio::fs::File, store: DB) -> anyhow::Result<Self> {
+        let roots = load_car_unchecked(&store, file.compat()).await?;
+
+        if roots.len() != 2 {
+            return Err(anyhow!("invalid v1 snapshot, should have 2 root cids"));
+        }
+
+        let state_tree_root = roots[0];
+        let block_state_params_cid = roots[1];
+
+        let block_state_params_bytes = if let Some(bytes) = store.get(&block_state_params_cid)? {
+            bytes
+        } else {
+            return Err(anyhow!(
+                "invalid v1 snapshot, block state params cid not found: {}",
+                block_state_params_cid
+            ));
+        };
+
+        Ok(Self {
+            state_tree_root,
+            state_tree: StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_tree_root)?,
+            block_state_params_bytes,
+            block_state_params_cid,
+        })
+    }
+
+    /// For V1 snapshot, we are putting two components into the CAR file: the state tree and latest
+    /// state params. One header is for the state tree root. The other header root cid is for the
+    /// serialized block state params.
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        Ok(vec![self.state_tree_root, self.block_state_params_cid])
+    }
+
+    pub fn init_chain(&self) -> anyhow::Result<()> {
+        todo!()
+    }
+
+    async fn write_car(self, car: CarHeader, write: tokio::fs::File) -> anyhow::Result<()> {
+        let write_task = tokio::spawn(async move {
+            let state_tree_streamer =
+                StateTreeStreamer::new(self.state_tree_root, self.state_tree.into_store());
+            let block_state_streamer = tokio_stream::iter(vec![(
+                self.block_state_params_cid,
+                self.block_state_params_bytes,
+            )]);
+            let mut streamer = state_tree_streamer.merge(block_state_streamer);
+
+            let mut write = write.compat_write();
+            car.write_stream_async(&mut Pin::new(&mut write), &mut streamer)
+                .await
+                .unwrap();
+        });
+
+        write_task.await?;
```

Comment:
Will a panic in the task be turned into an `Error`?

Reply:
returns the result now